### DISCUSSION TITLE
P1017 on prisma migrate - following Getting Started guide

### DISCUSSION THREAD

**Discussion by blueisgreen** [prisma/prisma#12972] [Q&A]
Just discovered Prisma, so I am giving it a try. Using the Getting Started guide (start from scratch, Postgres), the migration step fails with:

Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "my_db_name", schema "public" at "localhost:5432"

```
Error: P1017

Server has closed the connection.
```

I can connect via psql. DATABASE_URL works fine with my Knex setup. Is there a way to see detailed logging about what is going wrong?

**Comment by austincrim** [comment]
Hey @blueisgreen 👋 ,

You enable debug logging in the CLI commands by setting the `DEBUG` environment variable before running your commands, e.g. `DEBUG=* npx prisma migrate dev`.

Could you share your connection string (with sensitive info redacted)? Does your password have any special characters that need to be URL encoded?

**Comment by blueisgreen** [comment]
Thanks for helping. Here's the connection string, everything sensitive is changed but in a similar format (underscores).

DATABASE_URL="postgresql://dbuser:password@localhost:5432/db_name_dev1?schema=prisma"

"dbuser" is an admin. Works fine w/ knex migrations.

When I run w/ debugging, I see:

```
  prisma:migrateEngine:rpc SENDING RPC CALL {"id":1,"jsonrpc":"2.0","method":"devDiagnostic","params":{"migrationsDirectoryPath":"/project_path/prisma/migrations"}} +4ms
  prisma:migrateEngine:stderr {"timestamp":"2022-04-30T17:48:19.957258Z","level":"INFO","fields":{"message":"Starting migration engine RPC server","git_hash":"22b822189f46ef0dc5c5b503368d1bee01213980"},"target":"migration_engine"} +0ms
  prisma:migrateEngine:stderr {"timestamp":"2022-04-30T17:48:20.193373Z","level":"ERROR","fields":{"message":"Error in PostgreSQL connection: Error { kind: Closed, cause: None }"},"target":"quaint::connector::postgres"} +236ms
  prisma:migrateEngine:stderr {"timestamp":"2022-04-30T17:48:20.194643Z","level":"ERROR","fields":{"message":"Error in PostgreSQL connection: Error { kind: Io, cause: Some(Os { code: 54, kind: ConnectionReset, message: \"Connection reset by peer\" }) }"},"target":"quaint::connector::postgres"} +1ms
  prisma:migrateEngine:rpc {
  prisma:migrateEngine:rpc   jsonrpc: '2.0',
  prisma:migrateEngine:rpc   error: {
  prisma:migrateEngine:rpc     code: 4466,
  prisma:migrateEngine:rpc     message: 'An error happened. Check the data field for details.',
  prisma:migrateEngine:rpc     data: {
  prisma:migrateEngine:rpc       is_panic: false,
  prisma:migrateEngine:rpc       message: 'Server has closed the connection.',
  prisma:migrateEngine:rpc       meta: null,
  prisma:migrateEngine:rpc       error_code: 'P1017'
  prisma:migrateEngine:rpc     }
  prisma:migrateEngine:rpc   },
```

Also, a new, empty prisma_migrate_shadow_db... shows up, so it's at least connecting long enough to do that.

Any ideas? Kind of a non-starter.

**Reply by janpio** [reply]
> prisma:migrateEngine:stderr {"timestamp":"2022-04-30T17:48:20.193373Z","level":"ERROR","fields":{"message":"Error in PostgreSQL connection: Error { kind: Closed, cause: None }"},"target":"quaint::connector::postgres"} +236ms
>   prisma:migrateEngine:stderr {"timestamp":"2022-04-30T17:48:20.194643Z","level":"ERROR","fields":{"message":"Error in PostgreSQL connection: Error { kind: Io, cause: Some(Os { code: 54, kind: ConnectionReset, message: \"Connection reset by peer\" }) }"},"target":"quaint::connector::postgres"} +1ms

Can you check the database logs? For Prisma this is pretty much a "database does not want to talk to me any more" situation unfortunately and does (and can) not know more about this unfortunately.

**Comment by blueisgreen** [comment]
I wonder if it's because of this:

NOTICE:  identifier "pgAdmin 4 - DB:prisma_migrate_shadow_db_ade16be5-e2c3-4787-a0c2-b3059c7f935e" will be truncated to "pgAdmin 4 - DB:prisma_migrate_shadow_db_ade16be5-e2c3-4787-a0c2"

Maybe it's trying to connect with the full name, and that doesn't match. Just a guess.

**Reply by janpio** [reply]
That is at least an interesting observation. What are you using PgAdmin for in that contexthat is part of the identifier? Not sure how I can connect these dots.

**Comment by blueisgreen** [comment]
I assume it's related to something Prisma is doing. I am doing nothing special.

Anyway, clearly this won't work for me. I'm moving on.

**Reply by janpio** [reply]
Prisma creates a shadow database with the name `prisma_migrate_shadow_db_ade16be5-e2c3-4787-a0c2-b3059c7f935e` behind the scenes for `migrate dev`. Maybe you opened PgAdmin and looked at the created databases, and somehow that created an identifier also using the database name (maybe to identify the connection?) which needed to be truncated. (Can you confirm you were using PgAdmin to look at the databases?).

Still this is obviously totally not expected that the PostgreSQL server closes the connection :/

**Comment by storageddd** [comment]
Have the same issue. Just trying to add unique fields in my new migration, and after it, catch this error :(

**Reply by pantharshit00** [reply]
Hello @storageddd! Sorry for late follow up here.

If this is still relevant, can you let me where your database is hosted? Migrate needs permission to create a new temporary database in order to cleanly apply migrations so can you make sure your database user account has that permission?

**Comment by mausworks** [comment]
I had to create the shadow DB manually, then everything worked as expected.

Just write this in the Supabase SQL Editor:

```sql
CREATE DATABASE shadow;
```

_I know this is an old issue/thread, but I ended up here (again) from Google._

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
